File: bindings/rust/src/mcp.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//

//! MCP types for the proven-servers ABI.
//!
//! Formally verified Model Context Protocol types.
//! Mirrors the Idris2 module `McpABI.Types`.
//!
//! - `McpMessageType` -- MCP message types.
//! - `Transport` -- MCP transport types.
//! - `McpContentType` -- MCP content types.
//! - `McpErrorCode` -- MCP error codes.
//! - `McpCapability` -- MCP server capabilities.
//! - `SessionState` -- MCP session lifecycle states.
//!
//! All discriminant values match the Idris2 ABI tag definitions exactly.

use std::fmt;

// ===========================================================================
// McpMessageType (tags 0-13)
// ===========================================================================

/// MCP message types.
///
/// Matches `McpMessageType` in `McpABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum McpMessageType {
    /// Initialize (tag 0).
    Initialize = 0,
    /// Initialized (tag 1).
    Initialized = 1,
    /// Ping (tag 2).
    Ping = 2,
    /// CallTool (tag 3).
    CallTool = 3,
    /// ToolResult (tag 4).
    ToolResult = 4,
    /// ListTools (tag 5).
    ListTools = 5,
    /// ListResources (tag 6).
    ListResources = 6,
    /// ReadResource (tag 7).
    ReadResource = 7,
    /// ListPrompts (tag 8).
    ListPrompts = 8,
    /// GetPrompt (tag 9).
    GetPrompt = 9,
    /// Subscribe (tag 10).
    Subscribe = 10,
    /// Unsubscribe (tag 11).
    Unsubscribe = 11,
    /// Notification (tag 12).
    Notification = 12,
    /// Cancel (tag 13).
    Cancel = 13,
}

impl McpMessageType {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Initialize),
            1 => Some(Self::Initialized),
            2 => Some(Self::Ping),
            3 => Some(Self::CallTool),
            4 => Some(Self::ToolResult),
            5 => Some(Self::ListTools),
            6 => Some(Self::ListResources),
            7 => Some(Self::ReadResource),
            8 => Some(Self::ListPrompts),
            9 => Some(Self::GetPrompt),
            10 => Some(Self::Subscribe),
            11 => Some(Self::Unsubscribe),
            12 => Some(Self::Notification),
            13 => Some(Self::Cancel),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [McpMessageType; 14] = [
        Self::Initialize, Self::Initialized, Self::Ping, Self::CallTool, Self::ToolResult, Self::ListTools, Self::ListResources, Self::ReadResource, Self::ListPrompts, Self::GetPrompt, Self::Subscribe, Self::Unsubscribe, Self::Notification, Self::Cancel,
    ];
}

impl fmt::Display for McpMessageType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Transport (tags 0-3)
// ===========================================================================

/// MCP transport types.
///
/// Matches `Transport` in `McpABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum Transport {
    /// Stdio (tag 0).
    Stdio = 0,
    /// SSE (tag 1).
    Sse = 1,
    /// WebSocket (tag 2).
    WebSocket = 2,
    /// Streamable HTTP (tag 3).
    StreamableHttp = 3,
}

impl Transport {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Stdio),
            1 => Some(Self::Sse),
            2 => Some(Self::WebSocket),
            3 => Some(Self::StreamableHttp),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [Transport; 4] = [
        Self::Stdio, Self::Sse, Self::WebSocket, Self::StreamableHttp,
    ];
}

impl fmt::Display for Transport {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// McpContentType (tags 0-3)
// ===========================================================================

/// MCP content types.
///
/// Matches `McpContentType` in `McpABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum McpContentType {
    /// Text (tag 0).
    Text = 0,
    /// Image (tag 1).
    Image = 1,
    /// Resource (tag 2).
    Resource = 2,
    /// Embedding (tag 3).
    Embedding = 3,
}

impl McpContentType {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Text),
            1 => Some(Self::Image),
            2 => Some(Self::Resource),
            3 => Some(Self::Embedding),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [McpContentType; 4] = [
        Self::Text, Self::Image, Self::Resource, Self::Embedding,
    ];
}

impl fmt::Display for McpContentType {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// McpErrorCode (tags 0-5)
// ===========================================================================

/// MCP error codes.
///
/// Matches `McpErrorCode` in `McpABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum McpErrorCode {
    /// ParseError (tag 0).
    ParseError = 0,
    /// InvalidRequest (tag 1).
    InvalidRequest = 1,
    /// MethodNotFound (tag 2).
    MethodNotFound = 2,
    /// InvalidParams (tag 3).
    InvalidParams = 3,
    /// InternalError (tag 4).
    InternalError = 4,
    /// Timeout (tag 5).
    Timeout = 5,
}

impl McpErrorCode {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::ParseError),
            1 => Some(Self::InvalidRequest),
            2 => Some(Self::MethodNotFound),
            3 => Some(Self::InvalidParams),
            4 => Some(Self::InternalError),
            5 => Some(Self::Timeout),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [McpErrorCode; 6] = [
        Self::ParseError, Self::InvalidRequest, Self::MethodNotFound, Self::InvalidParams, Self::InternalError, Self::Timeout,
    ];
}

impl fmt::Display for McpErrorCode {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// McpCapability (tags 0-4)
// ===========================================================================

/// MCP server capabilities.
///
/// Matches `McpCapability` in `McpABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum McpCapability {
    /// Tools (tag 0).
    Tools = 0,
    /// Resources (tag 1).
    Resources = 1,
    /// Prompts (tag 2).
    Prompts = 2,
    /// Logging (tag 3).
    Logging = 3,
    /// Sampling (tag 4).
    Sampling = 4,
}

impl McpCapability {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Tools),
            1 => Some(Self::Resources),
            2 => Some(Self::Prompts),
            3 => Some(Self::Logging),
            4 => Some(Self::Sampling),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [McpCapability; 5] = [
        Self::Tools, Self::Resources, Self::Prompts, Self::Logging, Self::Sampling,
    ];
}

impl fmt::Display for McpCapability {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// SessionState (tags 0-4)
// ===========================================================================

/// MCP session lifecycle states.
///
/// Matches `SessionState` in `McpABI.Types`.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(u8)]
pub enum SessionState {
    /// Idle (tag 0).
    Idle = 0,
    /// Connecting (tag 1).
    Connecting = 1,
    /// Ready (tag 2).
    Ready = 2,
    /// Processing (tag 3).
    Processing = 3,
    /// Disconnecting (tag 4).
    Disconnecting = 4,
}

impl SessionState {
    /// Decode from an ABI tag value.
    pub fn from_tag(tag: u8) -> Option<Self> {
        match tag {
            0 => Some(Self::Idle),
            1 => Some(Self::Connecting),
            2 => Some(Self::Ready),
            3 => Some(Self::Processing),
            4 => Some(Self::Disconnecting),
            _ => None,
        }
    }

    /// Encode to the ABI tag value.
    pub fn to_tag(self) -> u8 {
        self as u8
    }

    /// All variants of this type.
    pub const ALL: [SessionState; 5] = [
        Self::Idle, Self::Connecting, Self::Ready, Self::Processing, Self::Disconnecting,
    ];
}

impl fmt::Display for SessionState {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        fmt::Debug::fmt(self, f)
    }
}

// ===========================================================================
// Tests
// ===========================================================================

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mcp_message_type_roundtrip() {
        for v in McpMessageType::ALL {
            let tag = v.to_tag();
            let decoded = McpMessageType::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(McpMessageType::from_tag(14).is_none());
    }

    #[test]
    fn transport_roundtrip() {
        for v in Transport::ALL {
            let tag = v.to_tag();
            let decoded = Transport::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(Transport::from_tag(4).is_none());
    }

    #[test]
    fn mcp_content_type_roundtrip() {
        for v in McpContentType::ALL {
            let tag = v.to_tag();
            let decoded = McpContentType::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(McpContentType::from_tag(4).is_none());
    }

    #[test]
    fn mcp_error_code_roundtrip() {
        for v in McpErrorCode::ALL {
            let tag = v.to_tag();
            let decoded = McpErrorCode::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(McpErrorCode::from_tag(6).is_none());
    }

    #[test]
    fn mcp_capability_roundtrip() {
        for v in McpCapability::ALL {
            let tag = v.to_tag();
            let decoded = McpCapability::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(McpCapability::from_tag(5).is_none());
    }

    #[test]
    fn session_state_roundtrip() {
        for v in SessionState::ALL {
            let tag = v.to_tag();
            let decoded = SessionState::from_tag(tag).expect("valid tag");
            assert_eq!(decoded, v);
        }
        assert!(SessionState::from_tag(5).is_none());
    }

}

File: protocols/proven-mcp/ffi/zig/src/mcp.zig
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// mcp.zig -- Zig FFI implementation of proven-mcp.
//
// Implements the MCP (Model Context Protocol) server state machine with:
//   - 64-slot mutex-protected session pool
//   - Per-session capability bitmask (Tools, Resources, Prompts, Logging, Sampling)
//   - Per-session pending request tracking (max 32 in-flight)
//   - Transport selection per session (Stdio, SSE, WebSocket, StreamableHTTP)
//   - Thread-safe via per-pool mutex
//
// All exported functions use C calling convention (callconv(.c)) and
// communicate state via u8 tags matching abi.Types.idr exactly.

const std = @import("std");

// =========================================================================
// Enums (matching abi.Types.idr tag assignments)
// =========================================================================

/// MCP message types (ABI tags 0-13).
pub const MessageType = enum(u8) {
    initialize = 0,
    initialized = 1,
    ping = 2,
    call_tool = 3,
    tool_result = 4,
    list_tools = 5,
    list_resources = 6,
    read_resource = 7,
    list_prompts = 8,
    get_prompt = 9,
    subscribe = 10,
    unsubscribe = 11,
    notification = 12,
    cancel = 13,
};

/// MCP transport types (ABI tags 0-3).
pub const Transport = enum(u8) {
    stdio = 0,
    sse = 1,
    websocket = 2,
    streamable_http = 3,
};

/// MCP content types (ABI tags 0-3).
pub const ContentType = enum(u8) {
    text = 0,
    image = 1,
    resource = 2,
    embedding = 3,
};

/// MCP error codes (ABI tags 0-5).
pub const ErrorCode = enum(u8) {
    parse_error = 0,
    invalid_request = 1,
    method_not_found = 2,
    invalid_params = 3,
    internal_error = 4,
    timeout = 5,
};

/// MCP server capabilities (ABI tags 0-4).
pub const Capability = enum(u8) {
    tools = 0,
    resources = 1,
    prompts = 2,
    logging = 3,
    sampling = 4,
};

/// MCP session lifecycle states (ABI tags 0-4).
pub const SessionState = enum(u8) {
    idle = 0,
    connecting = 1,
    ready = 2,
    processing = 3,
    disconnecting = 4,
};

// =========================================================================
// Internal data structures
// =========================================================================

/// Maximum concurrent sessions.
const MAX_SESSIONS: usize = 64;

/// Maximum pending requests per session.
const MAX_PENDING: usize = 32;

/// Maximum name length in bytes.
const MAX_NAME_LEN: usize = 256;

/// Maximum server name length.
const MAX_SERVER_NAME_LEN: usize = 128;

/// Number of capability bits.
const NUM_CAPABILITIES: usize = 5;

/// A pending request entry.
const PendingRequest = struct {
    /// Request ID.
    req_id: u32,
    /// Tool/method name.
    name: [MAX_NAME_LEN]u8,
    name_len: u32,
    /// Whether this slot is active.
    active: bool,
};

/// Default (empty) pending request.
const empty_pending: PendingRequest = .{
    .req_id = 0,
    .name = [_]u8{0} ** MAX_NAME_LEN,
    .name_len = 0,
    .active = false,
};

/// An MCP server session.
const Session = struct {
    /// Current session lifecycle state.
    state: SessionState,
    /// Transport type.
    transport: Transport,
    /// Server name.
    server_name: [MAX_SERVER_NAME_LEN]u8,
    server_name_len: u32,
    /// Capability bitmask (bit N = capability N enabled).
    capabilities: u8,
    /// Pending requests.
    pending: [MAX_PENDING]PendingRequest,
    /// Number of active pending requests.
    pending_count: u32,
    /// Whether this session slot is in use.
    active: bool,
};

/// Default (empty) session.
const empty_session: Session = .{
    .state = .idle,
    .transport = .stdio,
    .server_name = [_]u8{0} ** MAX_SERVER_NAME_LEN,
    .server_name_len = 0,
    .capabilities = 0,
    .pending = [_]PendingRequest{empty_pending} ** MAX_PENDING,
    .pending_count = 0,
    .active = false,
};

// =========================================================================
// Global state
// =========================================================================

var sessions: [MAX_SESSIONS]Session = [_]Session{empty_session} ** MAX_SESSIONS;
var mutex: std.Thread.Mutex = .{};

// =========================================================================
// Internal helpers
// =========================================================================

/// Validate a slot index, returning null if out of range or inactive.
fn validSlot(slot: c_int) ?usize {
    if (slot < 0 or slot >= MAX_SESSIONS) return null;
    const idx: usize = @intCast(slot);
    if (!sessions[idx].active) return null;
    return idx;
}

// =========================================================================
// Exported C ABI functions
// =========================================================================

/// Returns the ABI version number.
pub export fn mcp_abi_version() callconv(.c) u32 {
    return 1;
}

/// Create a new MCP session. Returns slot index (>=0) or -1 on failure.
pub export fn mcp_create(
    transport: u8,
    name_ptr: [*]const u8,
    name_len: u32,
) callconv(.c) c_int {
    mutex.lock();
    defer mutex.unlock();

    if (transport > 3) return -1;
    if (name_len == 0 or name_len > MAX_SERVER_NAME_LEN) return -1;

    for (&sessions, 0..) |*s, i| {
        if (!s.active) {
            s.* = empty_session;
            s.transport = @enumFromInt(transport);
            @memcpy(s.server_name[0..name_len], name_ptr[0..name_len]);
            s.server_name_len = name_len;
            s.state = .connecting; // Idle -> Connecting
            s.active = true;
            return @intCast(i);
        }
    }
    return -1;
}

/// Destroy a session, releasing its slot.
pub export fn mcp_destroy(slot: c_int) callconv(.c) void {
    mutex.lock();
    defer mutex.unlock();
    if (slot < 0 or slot >= MAX_SESSIONS) return;
    sessions[@intCast(slot)] = empty_session;
}

/// Returns the current SessionState tag for a session.
pub export fn mcp_state(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].state);
}

/// Initialize session with capabilities bitmask. Transitions Connecting -> Ready.
pub export fn mcp_initialize(slot: c_int, caps_bitmask: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .connecting) return 1;

    sessions[idx].capabilities = caps_bitmask;
    sessions[idx].state = .ready;
    return 0;
}

/// Add a single capability. Returns 0 on success, 1 on rejection.
pub export fn mcp_add_capability(slot: c_int, cap: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (cap >= NUM_CAPABILITIES) return 1;

    sessions[idx].capabilities |= (@as(u8, 1) << @intCast(cap));
    return 0;
}

/// Check if a capability is enabled. Returns 1 if yes, 0 if no.
pub export fn mcp_has_capability(slot: c_int, cap: u8) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 0;
    if (cap >= NUM_CAPABILITIES) return 0;

    return if ((sessions[idx].capabilities & (@as(u8, 1) << @intCast(cap))) != 0) 1 else 0;
}

/// Call a tool. Transitions Ready -> Processing. Returns 0 on success.
pub export fn mcp_call_tool(
    slot: c_int,
    name_ptr: [*]const u8,
    name_len: u32,
    req_id: u32,
) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state != .ready and state != .processing) return 1;
    if (name_len == 0 or name_len > MAX_NAME_LEN) return 1;
    if (sessions[idx].pending_count >= MAX_PENDING) return 1;

    // Find a free pending slot
    for (&sessions[idx].pending) |*p| {
        if (!p.active) {
            p.req_id = req_id;
            @memcpy(p.name[0..name_len], name_ptr[0..name_len]);
            p.name_len = name_len;
            p.active = true;
            sessions[idx].pending_count += 1;
            sessions[idx].state = .processing;
            return 0;
        }
    }
    return 1;
}

/// Complete a pending request. May transition Processing -> Ready.
pub export fn mcp_complete_request(slot: c_int, req_id: u32) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;

    for (&sessions[idx].pending) |*p| {
        if (p.active and p.req_id == req_id) {
            p.active = false;
            p.name_len = 0;
            sessions[idx].pending_count -= 1;

            if (sessions[idx].pending_count == 0 and
                sessions[idx].state == .processing)
            {
                sessions[idx].state = .ready;
            }
            return 0;
        }
    }
    return 1;
}

/// Cancel a pending request. Returns 0 on success, 1 on rejection.
pub export fn mcp_cancel_request(slot: c_int, req_id: u32) callconv(.c) u8 {
    // Same logic as complete — just removes the pending entry
    return mcp_complete_request(slot, req_id);
}

/// Returns the number of pending requests.
pub export fn mcp_pending_count(slot: c_int) callconv(.c) u32 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return sessions[idx].pending_count;
}

/// Disconnect the session. Returns 0 on success, 1 on rejection.
pub export fn mcp_disconnect(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state == .connecting or state == .ready or state == .processing) {
        sessions[idx].state = .disconnecting;
        return 0;
    }
    return 1;
}

/// Complete cleanup. Transitions Disconnecting -> Idle.
pub export fn mcp_cleanup(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    if (sessions[idx].state != .disconnecting) return 1;

    sessions[idx].state = .idle;
    sessions[idx].pending = [_]PendingRequest{empty_pending} ** MAX_PENDING;
    sessions[idx].pending_count = 0;
    sessions[idx].capabilities = 0;

    return 0;
}

/// Returns the transport tag for a session.
pub export fn mcp_transport(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();
    const idx = validSlot(slot) orelse return 0;
    return @intFromEnum(sessions[idx].transport);
}

/// Send a keepalive ping. Only valid from Ready or Processing.
pub export fn mcp_ping(slot: c_int) callconv(.c) u8 {
    mutex.lock();
    defer mutex.unlock();

    const idx = validSlot(slot) orelse return 1;
    const state = sessions[idx].state;
    if (state != .ready and state != .processing) return 1;
    return 0;
}

/// Check if a session state transition is valid.
pub export fn mcp_can_transition(from: u8, to: u8) callconv(.c) u8 {
    if (from == 0 and to == 1) return 1; // Idle -> Connecting
    if (from == 1 and to == 2) return 1; // Connecting -> Ready
    if (from == 2 and to == 3) return 1; // Ready -> Processing
    if (from == 3 and to == 3) return 1; // Processing -> Processing
    if (from == 3 and to == 2) return 1; // Processing -> Ready
    if (from == 1 and to == 4) return 1; // Connecting -> Disconnecting
    if (from == 2 and to == 4) return 1; // Ready -> Disconnecting
    if (from == 3 and to == 4) return 1; // Processing -> Disconnecting
    if (from == 4 and to == 0) return 1; // Disconnecting -> Idle
    return 0;
}
